casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -0,0 +1,280 @@
+import cats.syntax.all._
+import coop.rchain.blockstorage.BlockStore
+import coop.rchain.blockstorage.BlockStore.BlockStore
+import coop.rchain.blockstorage.dag.BlockDagStorage
+import coop.rchain.casper.engine.BlockRetriever
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.casper.syntax._
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.casper.{CasperShardConf, PrettyPrinter, Validate}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+import fs2.concurrent.Queue
+
+sealed trait RecvStatus
+// Checking if block should be ignored or continue with processing
+case object BeginReceived extends RecvStatus
+// Block received and stored in blocks store, waiting for validation and DAG insertion
+case object ReceivedAndStored extends RecvStatus
+// Block sent to validation
+case object PendingValidation extends RecvStatus
+// Requested missing dependencies
+case object PendingRequest extends RecvStatus
+
+object BlockReceiverState {
+  def apply[MId]: BlockReceiverState[MId] = new BlockReceiverState(
+    blocksSt = Map(),
+    receiveSt = Map(),
+    childRelations = Map()
+  )
+}
+
+final case class BlockReceiverState[MId](
+    /**
+      * Blocks received and stored in BlockStore with parent relations
+      */
+    blocksSt: Map[MId, Set[MId]],
+    /**
+      * Blocks receiving status
+      */
+    receiveSt: Map[MId, RecvStatus],
+    /**
+      * Blocks mapping with children relations
+      */
+    childRelations: Map[MId, Set[MId]]
+) {
+
+  /**
+    * Keep status of block as received process started
+    *  - thread safe sync like opening transaction for a block by block hash
+    */
+  def beginReceived(id: MId): (BlockReceiverState[MId], Boolean) = {
+    val expectedReceive = receiveSt.get(id).collect { case PendingRequest => true }.getOrElse(true)
+    // If state is not known or pending it's request, continue with receiving
+    if (expectedReceive) {
+      // Update state to begin received status
+      val newReceiveSt = receiveSt + ((id, BeginReceived))
+      (copy(receiveSt = newReceiveSt), true)
+    } else {
+      (this, false)
+    }
+  }
+
+  /**
+    * End of block receiving status update
+    *  - like closing transaction, block received and stored
+    *
+    *  @return unseen parent dependencies
+    */
+  def endReceived(id: MId, parents: List[(MId, Boolean)]): (BlockReceiverState[MId], Set[MId]) = {
+    val curStateOpt = receiveSt.get(id)
+    assert(
+      curStateOpt == BeginReceived.some,
+      s"Received should be called only in begin received state, actual: $curStateOpt, hash: $id"
+    )
+    curStateOpt
+      .collect {
+        case BeginReceived =>
+          // Update blocks state, keep unseen parents only
+          val parentsNotStored = parents.filter(_._2).map(_._1).toSet
+          val unseenParents    = parentsNotStored -- blocksSt.keySet -- receiveSt.keySet - id
```

Comment:
Right, `blocksSt.keySet` is needed to filter existing blocks already stored which can be as a justification of received block, `receiveSt` to filter the rest that are about to be stored and `id` is to eliminate strange probability `parents = [..., id]` with recursive definition.

Reply:
I still don't get it.
It is possible for block to end up in `blocksSt ` **only** after it is stored, because it is added only in `endStored` method. Therefore  if `parentsNotStored` contain contain some block which is a key in blocksSt - it should be a bug. Do I miss something?

How come `parents = [..., id]` is possible? This should be a bug as well if this happens?

Reply:
If blocks are received in this order. Critical is timing when 2. and 3. happens in the same time.
1. `b8{b5, b4}`
2. `b4{b2, b1}`  in the middle between updated state and before putting to BlockStore
3. `b6{b4, b3}`  b6 will not see b4 with BlockStore.contains but it will be in the state

---

> How come parents = [..., id] is possible? This should be a bug as well if this happens?

What you want to do if some _crazy_ block is received with self loop justification? Put invalid data in the state or ignore it?

Reply:
```
b4{b2, b1} in the middle between updated state and before putting to BlockStore
b6{b4, b3} b6 will not see b4 with BlockStore.contains but it will be in the state
```
But b4 will be in the state only after it is in BlockStore, it is added in `endStored` method which is supposed to be called only after block is stored.

```
What you want to do if some crazy block is received with self loop justification? Put invalid data in the state or ignore it?
```
I see, makes sense - maybe worth adding a comment. We want such blocks to be validated and sender slashed if this happens, so it should go through block reciever.

Reply:
Sorry, i didn't specify the case completely correct.

If `b6` executes this code - `b4` is not seen.
https://github.com/rchain/rchain/blob/d309b05c966c0de73896e013685399643fb68ffa/casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala#L246-L252

And then `b4` executes this - `b4` is now in the `blocksSt`.
https://github.com/rchain/rchain/blob/d309b05c966c0de73896e013685399643fb68ffa/casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala#L246-L253

Now `b6` executes this with `b4` not seen and state with `b4` in `blocksSt`.
https://github.com/rchain/rchain/blob/d309b05c966c0de73896e013685399643fb68ffa/casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala#L253

As we discussing on standup let's polish the code in future PRs with added tests.